### PR TITLE
New sample code - daemon calls protected web api

### PR DESCRIPTION
**Pull request by TiagoBrenck**
## Purpose
Show how a daemon application can call a protected web api using MSAL and making usage of app roles.

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* You can understand the steps on README file
* Your daemon application can call a web api

### CODE REVIEW

File: 2-Call-OwnApi/AppCreationScripts-withCert/sample.json [hunk 1]
```diff
@@ -0,0 +1,99 @@
+{
+  "Sample": {
+    "Title": "A .NET Core 2.x simple daemon console application calling a custom Web API with its own identity",
+    "Level": 200,
+    "Client": ".NET Core (Console)",
+    "Service": ".NET Core Web API",
+    "RepositoryUrl": "active-directory-dotnetcore-daemon-v2",
+    "Endpoint": "AAD v2.0"
```
Comment (jmprieur) [inline 2-Call-OwnApi/AppCreationScripts-withCert/sample.json]: Microsoft identity platform (we need to change the generator, BTW)

File: 2-Call-OwnApi/README.md [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+## Scenario
+
+The console application:
+
+- gets a token from Azure AD in its own name (without a user)
+- and then calls `TodoList-WebApi` endpoint to get the a list of todo's, displaying the result
+
+![Topology](./ReadmeFiles/daemon-with-secret.svg)
+
+For more information on the concepts used in this sample, be sure to read the [Microsoft identity platform endpoint client credentials protocol documentation](https://azure.microsoft.com/documentation/articles/active-directory-v2-protocols-oauth-client-creds).
```
Comment (jmprieur) [inline 2-Call-OwnApi/README.md]: We probably rather want to point customers to the daemon scenario landing page (which links to the protocol) ?
https://docs.microsoft.com/en-us/azure/active-directory/develop/scenario-daemon-overview

File: 2-Call-OwnApi/TodoList-WebApi/Startup.cs [hunk 3]
```diff
@@ -25,6 +29,28 @@ public Startup(IConfiguration configuration)
...
+            JwtSecurityTokenHandler.DefaultMapInboundClaims = false;
+
+            services.AddProtectWebApiWithMicrosoftIdentityPlatformV2(Configuration);
+
+            services.Configure<JwtBearerOptions>(AzureADDefaults.JwtBearerAuthenticationScheme, options =>
+            {
+                options.Events.OnTokenValidated = async context =>
+                {
+                    // This check is required to ensure that the Web API only accepts tokens from tenants where it has been consented and provisioned.
+                    if (!context.Principal.Claims.Any(y => y.Type == ClaimConstants.Roles && y.Value == "DaemonAppRole"))
```
Comment (jmprieur) [inline 2-Call-OwnApi/TodoList-WebApi/Startup.cs]: Don't we want to rather use an attribute on each action? As written it's correct, but it does not allow different controller actions to be protected with different app permissions / roles

File: Microsoft.Identity.Web/WebApiStartupHelpers.cs [hunk 4]
```diff
@@ -84,16 +81,16 @@ public static IServiceCollection AddProtectWebApiWithMicrosoftIdentityPlatformV2
...
-                          && !context.Principal.Claims.Any(y => y.Type == ClaimConstants.Roles))
-                       {
-                           throw new UnauthorizedAccessException("Neither scope or roles claim was found in the bearer token.");
-                       }
-
-                       await Task.FromResult(0);
-                   };
+                {
+                    // This check is required to ensure that the Web API only accepts tokens from tenants where it has been consented and provisioned.
+                    if (!context.Principal.Claims.Any(x => x.Type == ClaimConstants.Scope)
```
Comment (jmprieur) [inline Microsoft.Identity.Web/WebApiStartupHelpers.cs]: Why is this file needed given that you have already done the work in Startup.cs ?

Comment (TiagoBrenck) [reply Microsoft.Identity.Web/WebApiStartupHelpers.cs]: The `Startup` calls `AddProtectWebApiWithMicrosoftIdentityPlatformV2(Configuration);` first.
About the files, I have actually copied the whole `Microsoft.Identity.Web` project from our web app tutorial samples. Should I delete all the files that wont be used?

File: 2-Call-OwnApi/README.md (hunk 2 above)
Comment (TiagoBrenck) [reply 2-Call-OwnApi/README.md]: Changed

File: 2-Call-OwnApi/TodoList-WebApi/Startup.cs (hunk 3 above)
Comment (TiagoBrenck) [reply 2-Call-OwnApi/TodoList-WebApi/Startup.cs]: You re right. I have changed it and updated the readme

File: Microsoft.Identity.Web/WebApiStartupHelpers.cs (hunk 4 above)
Comment (jmprieur) [reply Microsoft.Identity.Web/WebApiStartupHelpers.cs]: no  @TiagoBrenck : don't remove the unused files. It's better to use the library entierely.

File: 2-Call-OwnApi/README.md [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+  - csharp
+  - azurepowershell
+products:
+  - azure-active-directory
+  - dotnet-core
+  - office-ms-graph
+description: "Shows how a daemon console app uses MSAL.NET to get an access token and call a Web API."
+---
+
+# A .NET Core 2.2 simple daemon console application calling a Web API with its own identity
```
Comment (kalyankrishna1) [inline 2-Call-OwnApi/README.md]: > 2.2 [](start = 13, length = 4)

remove versions. Its a liability to keep updating them

File: 2-Call-OwnApi/README.md [hunk 6]
```diff
@@ -0,0 +1,452 @@
...
+
+[![Build status](https://identitydivision.visualstudio.com/IDDP/_apis/build/status/AAD%20Samples/.NET%20client%20samples/active-directory-dotnetcore-daemon-v2%20CI)](https://identitydivision.visualstudio.com/IDDP/_build/latest?definitionId=695)
+
+## About this sample
+
+### Overview
+
+This sample application shows how to use the [Microsoft identity platform endpoint](http://aka.ms/aadv2) to access the data from a protected Web API, in a non-interactive process.  It uses the [OAuth 2 client credentials grant](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-client-creds-grant-flow) to acquire an access token, which can be used to call your own Web APIs. Additionally, it lays down all the steps developers need to take to secure their Web APIs with the Microsoft identity platform.
```
Comment (kalyankrishna1) [inline 2-Call-OwnApi/README.md]: >endpoint [](start = 74, length = 8)

remove

Comment (kalyankrishna1) [inline 2-Call-OwnApi/README.md]: >can be [](start = 361, length = 6)

is then

File: 2-Call-OwnApi/README.md [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+
+The app is a .NET Core 2.2 console application. It gets the list of "todos" from `TodoList-WebApi` project by using `Microsoft Authentication Library (MSAL) for .NET` to acquire a token.
```
Comment (kalyankrishna1) [inline 2-Call-OwnApi/README.md]: >Microsoft Authentication Library (MSAL) for .NET [](start = 117, length = 48)

hyperlink

Comment (kalyankrishna1) [inline 2-Call-OwnApi/README.md]: >a token. [](start = 178, length = 8)

an access token for the 'TodoList-WebApi'

File: 2-Call-OwnApi/README.md [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+
+## Scenario
+
+The console application:
+
+- gets a token from Azure AD in its own name (without a user)
```
Comment (kalyankrishna1) [inline 2-Call-OwnApi/README.md]: >gets a token from Azure AD in its own name (without a user) [](start = 2, length = 59)

Acquires an access token from the Microsoft Identity Platform as an application (no user interaction required)

File: 2-Call-OwnApi/README.md [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+- and then calls `TodoList-WebApi` endpoint to get the a list of todo's, displaying the result
```
Comment (kalyankrishna1) [inline 2-Call-OwnApi/README.md]: >endpoint [](start = 35, length = 8)

remove

File: 2-Call-OwnApi/README.md [hunk 10]
```diff
@@ -0,0 +1,452 @@
...
+
+![Topology](./ReadmeFiles/daemon-with-secret.svg)
+
+For more information on the concepts used in this sample, be sure to read the [Daemon application that calls web APIs documentation](https://docs.microsoft.com/en-us/azure/active-directory/develop/scenario-daemon-overview).
+
+> ### Daemon applications can use two forms of secrets to authenticate themselves with Azure AD:
+>
+> - **application secrets** (also named application password).
```
Comment (kalyankrishna1) [inline 2-Call-OwnApi/README.md]: >named [](start = 34, length = 5)

called

File: 2-Call-OwnApi/README.md [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+> - **certificates**.
+>
+> The first form (application secret) is treated in the next paragraphs.
+> A variation of this sample using a **certificate** instead, is available at the end of this article in [Variation: daemon application using client credentials with certificates](#Variation-daemon-application-using-client-credentials-with-certificates)
```
Comment (kalyankrishna1) [inline 2-Call-OwnApi/README.md]: >using [](start = 29, length = 5)

that uses

File: 2-Call-OwnApi/README.md [hunk 12]
```diff
@@ -0,0 +1,452 @@
...
+![Topology](./ReadmeFiles/daemon-with-secret.svg)
+
+For more information on the concepts used in this sample, be sure to read the [Daemon application that calls web APIs documentation](https://docs.microsoft.com/en-us/azure/active-directory/develop/scenario-daemon-overview).
+
+> ### Daemon applications can use two forms of secrets to authenticate themselves with Azure AD:
+>
+> - **application secrets** (also named application password).
+> - **certificates**.
+>
+> The first form (application secret) is treated in the next paragraphs.
```
Comment (kalyankrishna1) [inline 2-Call-OwnApi/README.md]: >form [](start = 12, length = 4)

approach

Comment (kalyankrishna1) [inline 2-Call-OwnApi/README.md]: >treated [](start = 41, length = 7)

discussed

File: 2-Call-OwnApi/README.md [hunk 13]
```diff
@@ -0,0 +1,452 @@
...
+Start the application, it will display the users in the tenant.
+
+## About the code
+
+The relevant code for this sample is in the `Program.cs` file, in the `RunAsync()` method. The steps are:
+
+1. Create the MSAL confidential client application.
+
+    Important note: even if we are building a console application, it is a daemon, and therefore a confidential client application, as it does not
+    access Web APIs on behalf of a user, but on its own application behalf.
```
Comment (kalyankrishna1) [inline 2-Call-OwnApi/README.md]: >application behalf [](start = 56, length = 18)

as an application

File: 2-Call-OwnApi/README.md [hunk 14]
```diff
@@ -0,0 +1,452 @@
...
+
+        await Task.FromResult(0);
+    };
+    ```
+
+    The protection can also be done on the `Controller` level, using the `Authorize` attribute:
+
+    ```csharp
+    [HttpGet]
+    [Authorize(Roles = "DaemonAppRole")]
```
Comment (kalyankrishna1) [inline 2-Call-OwnApi/README.md]: This is redundant. We checked the presence of roles earlier

File: 2-Call-OwnApi/README.md [hunk 15]
```diff
@@ -0,0 +1,452 @@
...
+## Variation: daemon application using client credentials with certificates
+
+Daemon applications can use two forms of secrets to authenticate themselves with Azure AD:
+
+- **application secrets** (also named application password). This is what we've seen so far.
+- **certificates**. This is the object of this paragraph.
+
+![Topology](./ReadmeFiles/daemon-with-certificate.svg)
+
+To use certificates instead of an application secret you will need to do little changes to what you have done so far:
```
Comment (kalyankrishna1) [inline 2-Call-OwnApi/README.md]: >use [](start = 3, length = 3)

Provide a link to the Client credentials protocol flow with certificates ([https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-client-creds-grant-flow#second-case-access-token-request-with-a-certificate](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-client-creds-grant-flow#second-case-access-token-request-with-a-certificate))

File: 2-Call-OwnApi/AppCreationScripts/sample.json [hunk 16]
```diff
@@ -0,0 +1,99 @@
+{
+  "Sample": {
+    "Title": "A .NET Core 2.x simple daemon console application calling a custom Web API with its own identity",
+    "Level": 200,
+    "Client": ".NET Core (Console)",
+    "Service": ".NET Core Web API",
+    "RepositoryUrl": "active-directory-dotnetcore-daemon-v2",
+    "Endpoint": "AAD v2.0"
```
Comment (kalyankrishna1) [inline 2-Call-OwnApi/AppCreationScripts/sample.json]: >AAD v2.0 [](start = 17, length = 8)

Microsoft Identity platform

File: 2-Call-OwnApi/AppCreationScripts/sample.json [hunk 17]
```diff
@@ -0,0 +1,99 @@
+{
+  "Sample": {
+    "Title": "A .NET Core 2.x simple daemon console application calling a custom Web API with its own identity",
```
Comment (kalyankrishna1) [inline 2-Call-OwnApi/AppCreationScripts/sample.json]: >2.x [](start = 26, length = 3)

remove versions

File: 2-Call-OwnApi/AppCreationScripts-withCert/AppCreationScripts.md [hunk 18]
```diff
@@ -0,0 +1,146 @@
+# Registering the Azure Active Directory applications and updating the configuration files for this sample using PowerShell scripts
```
Comment (kalyankrishna1) [inline 2-Call-OwnApi/AppCreationScripts-withCert/AppCreationScripts.md]: >Registering the Azure Active Directory applications and updating the configuration files [](start = 2, length = 88)

change the code gen for this "Register the sample apps with Microsoft Identity Platform and update the configuration files"

File: 2-Call-OwnApi/daemon-console/appsettings.json [hunk 19]
```diff
@@ -1,7 +1,9 @@
...
-  "Instance": "https://login.microsoftonline.com/{0}",
-  "Tenant": "[Enter here the tenantID or domain name for your Azure AD tenant]",
-  "ClientId": "[Enter here the ClientId for your application]",
-  "ClientSecret": "[Enter here a client secret for your application]",
-  "CertificateName": "[Or instead of client secret: Enter here the name of a certificate (from the user cert store) as registered with your application]"
+	"Instance": "https://login.microsoftonline.com/{0}",
+	"Tenant": "[Enter here the tenantID or domain name for your Azure AD tenant]",
+	"ClientId": "[Enter here the ClientId for your application]",
+	"ClientSecret": "[Enter here a client secret for your application]",
+	"CertificateName": "[Or instead of client secret: Enter here the name of a certificate (from the user cert store) as registered with your application]",
```
Comment (kalyankrishna1) [inline 2-Call-OwnApi/daemon-console/appsettings.json]: >the name of a certificate  [](start = 62, length = 26)

Its not the "name", its either "subject name" or "issued to". Find out and use the right term. Will also need clarifying this in readme.md

File: 2-Call-OwnApi/TodoList-WebApi/appsettings.json [hunk 20]
```diff
@@ -1,8 +1,14 @@
...
-    "LogLevel": {
-      "Default": "Warning"
-    }
-  },
-  "AllowedHosts": "*"
+	"AzureAd": {
+		"Instance": "https://login.microsoftonline.com/",
+		"ClientId": "[Enter the Client Id of the service (Application ID obtained from the Azure portal), e.g. ba74781c2-53c2-442a-97c2-3d60re42f403]",
+		"Domain": "[Enter the domain of your tenant, e.g. contoso.onmicrosoft.com]",
+		"TenantId": "[Enter 'common', or 'organizations' or the Tenant Id (Obtained from the Azure portal. Select 'Endpoints' from the 'App registrations' blade and use the GUID in any of the URLs), e.g. da41245a5-11b3-996c-00a8-4d99re19f292]"
```
Comment (kalyankrishna1) [inline 2-Call-OwnApi/TodoList-WebApi/appsettings.json]: >organizations [](start = 36, length = 13)

Lets start providing a link to a page ([https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-v2-protocols#endpoints](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-v2-protocols#endpoints)) that explains this.

File: 2-Call-OwnApi/TodoList-WebApi/Startup.cs [hunk 21]
```diff
@@ -25,6 +30,29 @@ public Startup(IConfiguration configuration)
...
+                options.Events.OnTokenValidated = async context =>
+                {
+                    // This check is required to ensure that the Web API only accepts tokens from tenants where it has been consented and provisioned.
+                    if (!context.Principal.Claims.Any(y => y.Type == ClaimConstants.Roles))
+                    {
+                        throw new UnauthorizedAccessException("Role claim was not found in the bearer token.");
+                    }
+
+                    await Task.FromResult(0);
+                };
```
Comment (kalyankrishna1) [inline 2-Call-OwnApi/TodoList-WebApi/Startup.cs]: This is redundant, the common method does this already

File: 2-Call-OwnApi/README.md (hunk 14 above)
Comment (TiagoBrenck) [reply 2-Call-OwnApi/README.md]: @kalyankrishna1, Jean-Marc suggested to use this attribute on top of an action result because most of the real applications would have few roles and the developers would probably decorate each action separately. It is a common practice. Earlier, we check if there is any `role` claim, without specifying which one.
